### ISSUE TITLE
the message is lose when the client is closed with Shared subscriptions 

### ISSUE THREAD

**Issue by vul3yo6** [eclipse/mosquitto#2662]
I use the Mosquitto 2.0.15 on Windows 10, and i found a problem:

the message is losed when i publish it with QoS1, and then i try close one of the three clients.
finally no any client receive the message which should received by the close client.

> the MQTT spec:
If the Server is in the process of sending a QoS 1 message to its chosen subscribing Client and the connection to that Client breaks before the Server has received an acknowledgement from the Client, the Server MAY wait for the Client to reconnect and retransmit the message to that Client. **If the Client'sSession terminates before the Client reconnects, the Server SHOULD send the Application Message to another Client that is subscribed to the same Shared !.** It MAY attempt to send the message to another Client as soon as it loses its connection to the first Client.

[2022-10-20_Mosquitto_share_subscription_lose_message](https://user-images.githubusercontent.com/4761902/196876951-ee8d850e-1304-4708-adbb-ae97b0a81d16.png)

**Comment by Daedaluz** [comment]
adding a -c flag to the subscribers seem to work for me.
Maybe because -x doesn't imply a persistent client.

**Comment by vul3yo6** [comment]
Thank you, Daedaluz
But i still have some problems.

The Mosquitto will send the Application Message to another Client that is subscribed to the same Shared Subscription.
But it only work for the **new** messages, how about the **old** messages?

![mosquitto_shared分配方式_consumer於0秒後斷線_說明](https://user-images.githubusercontent.com/4761902/197427954-3b7d9c93-d11c-481f-b3d6-bbf47a839e49.png)

The sever lose the old messages when you setting -c (- x defaults to -1 if -c is also given)
「.\mosquitto_sub -h localhost -t '$share/group1/test/topic' -c -i avm_sub_1 -q 1 -V mqttv5」
![mosquitto_shared分配方式_consumer於0秒後斷線_舊訊息遺失_說明](https://user-images.githubusercontent.com/4761902/197428212-168c939c-51fd-4855-ae1f-e3977cc41a28.png)

and then we can look back the origin question, 
The Mosquitto will send the **new** messages to another Client that is subscribed to the same Shared Subscription,
But Mosquitto will **NOT** send the **old** messages to another Client that is subscribed to the same Shared Subscription

I have three requirement
1. the server needs queue the old messages
2. the server needs send the **old** Message to another Client that is subscribed to the same Shared Subscription, If the Client'sSession terminates before the Client reconnects.
3. the server needs send the **new** Message to another Client that is subscribed to the same Shared Subscription, If the Client'sSession terminates before the Client reconnects.

**Comment by Daedaluz** [comment]
Right, the already queued messages should be redistributed among the other shared subscribers upon session expiry or disconnect as per spec.

I can observe the same problem.

**Comment by kumboleijo** [comment]
I'm currently facing the same issue - any news on this?

**Comment by Daedaluz** [comment]
peeking among the develop branch commits, I can't really find any message that is obviously related to this issue.
Mostly things about leak fixes and "coverty scans" =(

**Comment by dberardo-com** [comment]
i was wondering ... if clients connect using clean_session = False (even with QoS=0) and a shared subscription. When one client disconnects, then the broker will still queue up messages waiting for this client to connect again right? 

is there a way to configure a sort of timeout from the client side so that the server removes the connection after it has expired (even on clean_session=false)  ?

**Comment by Daedaluz** [comment]
My guess is that it will queue up the message on a connected client if there are other subscribers and otherwise just drop it. QoS=0 is after all "fire and forget".

That said, i think there is a setting `queue_qos0_messages` that queues QoS=0 messages to persistent clients.
There is also a setting `persistent_client_expiration` to remove persistent clients after a given duration of inactivity.

Also, MQTTv5 should have some per-message / per-client timout properties you could set.
eg Session Expiry Interval and Message Expiry Interval.

**Comment by dberardo-com** [comment]
i have made some tests recently and it seems that this https://github.com/eclipse/mosquitto/issues/2662#issuecomment-1711149584 is indeed an issue of clean_session=false ... everyone should be aware of this, i dont know if there is a place on the internet to document this

**Comment by LiamDavey** [comment]
I have come across this too.

I had two MQTTv5 clients with a shared subscription, QoS > 1, and Session_Expiry_Interval = 2 hours. One of the clients went offline. Mosquitto continued to send only half the messages to the online client. I assume (I haven't double checked) the other half of messages are being queued for the offline client. Once 2 hours was reached the offline client's session expired and all the queued messages were lost. Then the online client began to receive all the messages (not half) again.

I think Mosquitto should not queue QoS > 0 msgs for offline clients that are part of a shared subscription. Instead Mosquitto should forward the msgs to one of the online shared subscribers. This appears to be how [other brokers](https://docs.hivemq.com/hivemq/3.4/user-guide/shared-subscriptions.html) function. _"Offline clients with persistent sessions are not considered for distributing in the shared subscription algorithm."_.

I suppose for now I will have to decide between using QoS = 1, or using shared subscriptions with QoS = 0.

**Comment by Daedaluz** [comment]
@ralight This needs another set of eyes

**Comment by guilhermecampello** [comment]
Any news on this issue? I'm facing the same problem here.

When application replicas subscribe to the same shared subscription—for load balancing or redundancy—if one of those clients goes offline, messages continue to be sent to that offline client. In many cases, those messages may never be delivered because the offline client will never be reachable as it is a temporary instance created by auto-scaling mechanisms.
